google/cloud/bigtable/data/_async/_read_rows.py

```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
+                table_name=table.table_name,
+                app_profile_id=table.app_profile_id,
+            )
+        else:
+            self.request = query._to_pb(table)
+        self.table = table
+        self._predicate = retries.if_exception_type(
             core_exceptions.DeadlineExceeded,
             core_exceptions.ServiceUnavailable,
             core_exceptions.Aborted,
         )
-
-        def on_error_fn(exc):
-            if predicate(exc):
-                self.transient_errors.append(exc)
-
-        retry = retries.AsyncRetry(
-            predicate=predicate,
-            timeout=self.operation_timeout,
-            initial=0.01,
-            multiplier=2,
-            maximum=60,
-            on_error=on_error_fn,
-            is_stream=True,
+        self._metadata = _make_metadata(
+            table.table_name,
+            table.app_profile_id,
         )
-        self._stream: AsyncGenerator[Row, None] | None = retry(
-            self._partial_retryable
-        )()
-        # contains the list of errors that were retried
-        self.transient_errors: List[Exception] = []
-
-    def __aiter__(self) -> AsyncIterator[Row]:
-        """Implements the AsyncIterable interface"""
-        return self
-
-    async def __anext__(self) -> Row:
-        """Implements the AsyncIterator interface"""
-        if self._stream is not None:
-            return await self._stream.__anext__()
-        else:
-            raise asyncio.InvalidStateError("stream is closed")
+        self._last_yielded_row_key: bytes | None = None
+        self._remaining_count = self.request.rows_limit or None
 
-    async def aclose(self):
-        """Close the stream and release resources"""
-        if self._stream is not None:
-            await self._stream.aclose()
-        self._stream = None
-        self._emitted_seen_row_key = None
+    async def start_operation(self) -> AsyncGenerator[Row, None]:
+        """
+        Start the read_rows operation, retrying on retryable errors.
+        """
+        transient_errors = []
 
-    async def _read_rows_retryable_attempt(
-        self,
-        gapic_fn: Callable[..., Awaitable[AsyncIterable[ReadRowsResponse]]],
-        timeout_generator: Iterator[float],
-        total_row_limit: int,
-    ) -> AsyncGenerator[Row, None]:
+        def on_error_fn(exc):
+            if self._predicate(exc):
+                transient_errors.append(exc)
+
+        retry_fn = AsyncRetryableGenerator(
+            self.read_rows_attempt,
+            self._predicate,
+            exponential_sleep_generator(0.01, 60, multiplier=2),
+            self.operation_timeout,
+            on_error_fn,
+        )
+        try:
+            async for row in retry_fn:
+                yield row
+                if self._remaining_count is not None:
+                    self._remaining_count -= 1
+                    if self._remaining_count < 0:
+                        raise RuntimeError("emit count exceeds row limit")
```

Comment:
Should this be a more specific error? Maybe add a todo to improve it in the future

Reply:
Sure, I opened an issue to track this here: https://github.com/googleapis/python-bigtable/issues/853